### PR TITLE
EXPORT MODEL

### PR DESCRIPTION
**Pull request by rajagurunath**
1. export ml model into pickle, joblib, mlflow .
2. Added some ml docs
3. Test cases for testing the same

### CODE REVIEW

File: dask_sql/physical/rel/custom/export_model.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        EXPORT MODEL <model_name> WTIH (
+            model_serde = "pickle",
+            location = "model.pkl"
+        )
+    1. Most of the framework support pickle as a serilization
+        example:
+            sklearn
+            Dask ML
+            Pytorch
+    2. To Export a Universal (Framework agnostic model) use mlflow format
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:25]:
```suggestion
    2. To export a universal (framework agnostic) model, use the mlflow (https://mlflow.org/) format.
```

File: dask_sql/physical/rel/custom/export_model.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        - A mlflow is a framework which supports different flavours of model serilization, implemented
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:26]:
```suggestion
        - mlflow is a framework, which supports different flavors of model serialization, implemented
```

File: dask_sql/physical/rel/custom/export_model.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        for different ML libraries.
+        - Mlflow model is a self contained directory full of files which helps to load the
+        saved model without any environment Error
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:29]:
```suggestion
        - A mlflow model is a self-contained artifact, which contains everything you need for loading the model - without import errors.
```

File: dask_sql/physical/rel/custom/export_model.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        - To reproduce the environment conda.yaml files are produced while saving the model
+        Files saved as part of mlflow model
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:31]:
```suggestion
        - To reproduce the environment, conda.yaml files are produced while saving the model and stored as part of the mlflow model
```

File: dask_sql/physical/rel/custom/export_model.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            - conda.yaml (packages required to import saved model)
+            - MLmodel (A artifact about model which contains following)
+                - mlflow function which can load this saved model
+                - model was saved timestmap  etc..
+            - model : a ml model itself
+                - sklearn - model.pkl
+                - xgboost - model.xgb
+                - lightgbm - model.lgb
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:39]: I think we should not reproduce the mlflow documentation here and also not the mlflow internals (e.g. what is stored and what is not). If this changes at some point (due to whatever reason), we would need to update our comment here.

I think for the user, they either know already about mlflow or they can have a look into the documentation using the URL above. I would rather include the information, which libraries need to be installed for using mlflow.

File: dask_sql/physical/rel/custom/export_model.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    3. Export into ONNX (open standard ) -
+         - Needs input column names and their type to serilizes the model
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:41]: It's good to have this as a reminder, that this might come in the future but I would not include this in the documentation (otherwise it looks as if we can do it already). If you want, you can include a `TODO`.

File: dask_sql/physical/rel/custom/export_model.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        model_name = str(sql.getModelName().getIdentifier())
+        kwargs = convert_sql_kwargs(sql.getKwargs())
+        model_serde = kwargs.pop("model_serde", "pickle")
+        location = kwargs.pop("location", "tmp.pkl")
+        try:
+            model, training_columns = context.models[model_name]
+        except KeyError:
+            raise RuntimeError(f"A model with the name {model_name} is not present.")
+
+        if model_serde.lower() == "pickle" or model_serde.lower() == "pkl":
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:60]: You could move the `.lower()` to the definition in line 53, so you do not need to repeat it all the time.
```suggestion
        if model_serde in ["pickle", "pkl"]:
```

File: dask_sql/physical/rel/custom/export_model.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            with open(location, "wb") as pkl_file:
+                pickle.dump(model, pkl_file)
+        elif model_serde.lower() == "joblib":
+            joblib.dump(model, location)
+        # somewhat Experimental
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:65]: Good you make this comment! It would be even better to include this in the documentation of the class and add a comment here, *why* it is still experimental

File: dask_sql/physical/rel/custom/export_model.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        elif model_serde.lower() == "mlflow":
+            try:
+                import lightgbm
+                import mlflow
+                import xgboost
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:70]: I am not an expert in `mlflow`, but do we really need all three libraries if we want to export a model? E.g. when we want to export a non-xgboost model, do we need to have `xgboost` installed?

If not, you can do something like (not for mlflow, but for the other libraries)

```python
try:
    import lightgbm
except ImportError:
    lightgbm = None
```

and then later 
```python
    elif lightgbm and isinstance(model, lightgbm.basic.Booster):
```
because either the user has `lightgbm` installed or not - and then they will not be able to create a `lightgbm` model anyways :-)

For `mlflow`, you can keep your current logic (because without mlflow, the mlflow format setting does not make sense and we can abort early).

File: dask_sql/physical/rel/custom/export_model.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+                )
+            if isinstance(model, sklearn.base.BaseEstimator):
+                mlflow.sklearn.save_model(model, location, **kwargs)
+            elif isinstance(model, xgboost.core.Booster):
+                mlflow.xgboost.save_model(model, location, **kwargs)
+            elif isinstance(model, lightgbm.basic.Booster):
+                mlflow.lightgbm.save_model(model, location, **kwargs)
+            else:
+                raise NotImplementedError(
+                    f"Model {model.__class__} was not implemented"
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:83]:
```suggestion
                    f"mlflow export of type {model.__class__} is not implemented (yet)"
```

Let's hope users will help us implementing more model types ;-)

File: docs/pages/machine_learning.rst [hunk 11]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
 using the given name.
 Have a look into :ref:`ml` for more information.
 
+4. Check Model parameters - Model meta data
+-------------------------------------------
+After the Model was trained, you can inspect and get model details by using following
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:91]:
```suggestion
After the model was trained, you can inspect and get model details by using the following
```

File: docs/pages/machine_learning.rst [hunk 12]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+sql statements
+
+.. code-block:: sql
+
+    -- show list of models  which are trained and saved in the context.
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:96]:
```suggestion
    -- show the list of models which are trained and stored in the context.
```

File: docs/pages/machine_learning.rst [hunk 13]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    SHOW MODELS
+
+    -- To get Hyperparameters of the trained MODEL use ,
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:99]:
```suggestion
    -- To get the hyperparameters of the trained model, use 
```

File: docs/pages/machine_learning.rst [hunk 14]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    -- DESCRIBE MODEL <model_name>.
+    DESCRIBE MODEL my_model
+
+
+5. Export Trained Model
+------------------------
+Once your model was trained and performs good in your validation dataset,
+you can export the model into a file with supporting model serilization
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:107]:
```suggestion
you can export the model into a file with one of the supported model serialization
```

File: docs/pages/machine_learning.rst [hunk 15]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+format like pickle,joblib,mlflow (framework agnostic serilization format) etc.
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:108]:
```suggestion
formats like pickle, joblib, mlflow (framework-agnostic serialization format), etc.
```

File: docs/pages/machine_learning.rst [hunk 16]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+
+Currently dask_sql supports pickle,joblib and mlflow
+formats for exporting the trained model which can be deployed as microservices etc
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:111]:
```suggestion
Currently, dask-sql supports the pickle, joblib and mlflow format for exporting the trained model, which can then be deployed as microservices etc
```

File: docs/pages/machine_learning.rst [hunk 17]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+
+Before training and exporting the models from different framework like
+lightgbm, catboost, please ensure the relevant packages are installed in the
+dask_sql environment otherwise it will raises RuntimeError and if you
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:115]:
```suggestion
dask-sql environment, otherwise it will raise an exception on import and if you
```

File: docs/pages/machine_learning.rst [hunk 18]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+are using mlflow as model_serde ensure mlflow was installed
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:116]: If you change the name of the parameter, also change it here :-)

File: dask_sql/physical/rel/custom/export_model.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+        elif model_serde.lower() == "joblib":
+            joblib.dump(model, location)
+        # somewhat Experimental
+        elif model_serde.lower() == "mlflow":
+            try:
+                import lightgbm
+                import mlflow
+                import xgboost
+            except ImportError as impError:
+                raise RuntimeError(
```
Comment (nils-braun) [inline dask_sql/physical/rel/custom/export_model.py:72]: Please keep raising an `ImportError`, because some libraries handle those differently than other exceptions (which is good, as they can autodetect some things).
If you want, you can add a more telling error message, such as "For export in the mlflow format, you need to have mlflow installed" etc. - but you can also keep it

File: docs/pages/machine_learning.rst [hunk 20]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+
+
+.. code-block:: sql
+
+
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:121]:
```suggestion

```

File: docs/pages/machine_learning.rst [hunk 21]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    -- for pickle model serilization
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:122]:
```suggestion
    -- for pickle model serialization
```

File: docs/pages/machine_learning.rst [hunk 22]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    EXPORT MODEL my_model WITH (
+        model_serde ='pickle',
+        location = 'model.pkl')
+
+    -- for joblib model serilization
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:127]:
```suggestion
    -- for joblib model serialization
```

File: docs/pages/machine_learning.rst [hunk 23]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    EXPORT MODEL my_model WITH (
+        model_serde ='joblib',
+        location = 'model.pkl')
+
+    -- for mlflow model serilization
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:132]:
```suggestion
    -- for mlflow model serialization
```

File: docs/pages/machine_learning.rst [hunk 24]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+    EXPORT MODEL my_model WITH (
+        model_serde ='mlflow',
+        location = 'mlflow_dir')
+
+    -- Note you can pass more number of key value pairs
+    -- (parameters) which will delegated to respective
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:138]:
```suggestion
    -- (parameters) which will be delegated to the respective
```

File: docs/pages/machine_learning.rst [hunk 25]
```diff
@@ -86,6 +86,59 @@ query can be given). The model can than be used in subsequent calls to ``PREDICT
...
+Currently dask_sql supports pickle,joblib and mlflow
+formats for exporting the trained model which can be deployed as microservices etc
+
+Before training and exporting the models from different framework like
+lightgbm, catboost, please ensure the relevant packages are installed in the
+dask_sql environment otherwise it will raises RuntimeError and if you
+are using mlflow as model_serde ensure mlflow was installed
+
+
+.. code-block:: sql
```
Comment (nils-braun) [inline docs/pages/machine_learning.rst:119]: Just a matter of taste, but I like the closing ")" on the next line better (as you have done further down), so:
```sql
EXPORT MODEL (
    ...
)
```

File: planner/src/main/codegen/includes/model.ftl [hunk 26]
```diff
@@ -42,6 +42,25 @@ SqlNode SqlPredictModel() :
...
+{
+    final Span s;
+    final SqlModelIdentifier modelName;
+    final SqlKwargs kwargs;
+
+}
+{
+    <EXPORT> { s = span(); }
+    modelName = ModelIdentifier()
+    (
```
Comment (nils-braun) [inline planner/src/main/codegen/includes/model.ftl:56]: I think you do not need those "(" and ")" here

File: tests/integration/test_model.py [hunk 27]
```diff
@@ -334,3 +338,73 @@ def test_describe_model(c, training_df):
 
     with pytest.raises(RuntimeError):
         c.sql("DESCRIBE MODEL undefined_model")
+
+
+def test_export_model(c, training_df, tmpdir):
+    with pytest.raises(RuntimeError):
```
Comment (nils-braun) [inline tests/integration/test_model.py:344]: Well done! You are testing the corner cases - really good!

File: tests/integration/test_model.py [hunk 28]
```diff
@@ -334,3 +338,73 @@ def test_describe_model(c, training_df):
...
+            target_column = 'target'
+        ) AS (
+            SELECT x, y, x*y > 0 AS target
+            FROM timeseries
+            LIMIT 100
+        )
+    """
+    )
+    # Happy flow
+    temporary_file = tmpdir + "/pickle_model.pkl"
```
Comment (nils-braun) [inline tests/integration/test_model.py:364]: You could use `os.path.join(tmpdir, "pickle_model.pkl")`, which works on operation systems (maybe need to import `os` before)

File: tests/integration/test_model.py [hunk 29]
```diff
@@ -334,3 +338,73 @@ def test_describe_model(c, training_df):
...
+        " location = '{}')".format(temporary_file)
+    )
+
+    assert (
+        joblib.load(str(temporary_file)).__class__.__name__
+        == "GradientBoostingClassifier"
+    )
+    try:
+        # Test only when mlflow was installed
+        import mlflow
```
Comment (nils-braun) [inline tests/integration/test_model.py:388]: In `pytest`, there is a decorator to make tests skip if a library is not installed. Have a look [here](https://www.scivision.dev/pytest-skip-importerror/).
I think that would be the cleanest solution. In this case, you would need to create a single function for only the mlflow test.

File: tests/integration/test_model.py [hunk 30]
```diff
@@ -334,3 +338,73 @@ def test_describe_model(c, training_df):
...
+        )
+        assert len(os.listdir(temporary_dir)) == 3
+        assert (
+            mlflow.sklearn.load_model(str(temporary_dir)).__class__.__name__
+            == "GradientBoostingClassifier"
+        )
+    except ImportError as e:
+        print(e)
+
+    with pytest.raises(NotImplementedError):
```
Comment (nils-braun) [inline tests/integration/test_model.py:404]: Good!

**Review by nils-braun — COMMENTED** [review]
Hi @rajagurunath 
Again, really good job. It took a bit longer for the review - sorry about this. I tried to also comment a lot on some quite small (and mostly not so relevant) issues on the documentation - so please don't be discouraged by the large number of comments - they are mostly very small.

The build ja still failing, but I did not have a look so far. Shall I do so?

File: dask_sql/context.py [hunk 31]
```diff
@@ -107,6 +107,7 @@ def __init__(self):
         RelConverter.add_plugin_class(custom.ShowTablesPlugin, replace=False)
         RelConverter.add_plugin_class(custom.ShowModelsPlugin, replace=False)
         RelConverter.add_plugin_class(custom.ShowModelParamsPlugin, replace=False)
+        RelConverter.add_plugin_class(custom.SqlExportModel, replace=False)
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/context.py]: Got it, updated this time 👍

File: dask_sql/physical/rel/custom/export_model.py [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+import sklearn
+
+from dask_sql.datacontainer import DataContainer
+from dask_sql.physical.rel.base import BaseRelPlugin
+from dask_sql.utils import convert_sql_kwargs
+
+
+class SqlExportModel(BaseRelPlugin):
+    """
+    Export a trained model into a file using supported model serilization or model serDe.
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+
+    Sql:
+        EXPORT MODEL <model_name> WTIH (
+            model_serde = "pickle",
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: That perfectly makes sense, agreed, we need to take care of user experience

File: dask_sql/physical/rel/custom/export_model.py [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+            location = "model.pkl"
+        )
+    1. Most of the framework support pickle as a serilization
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py [hunk 35]
```diff
@@ -0,0 +1,99 @@
...
+        example:
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py [hunk 36]
```diff
@@ -0,0 +1,99 @@
...
+            sklearn
+            Dask ML
```
[the remark this answers is not in the document]
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: got it, Thanks 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 1 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 2 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 3 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 4 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 5 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: This one also perfectly makes sense !! removed those comments 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 6 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: Added as a TODO

File: dask_sql/physical/rel/custom/export_model.py (hunk 7 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: got it , updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 8 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: I think, since now we are doing the import check for each library separately, I feel this mlfow export will work for most of the cases without breaking. Please let me know if you see any breaking point we will mark it as experimental

File: dask_sql/physical/rel/custom/export_model.py (hunk 9 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: This is exactly the trick I am searching for, Thanks a lot for this !! 💯

File: dask_sql/physical/rel/custom/export_model.py (hunk 10 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: updated 👍

File: docs/pages/machine_learning.rst (hunk 11 above)
Comment (rajagurunath) [reply docs/pages/machine_learning.rst]: updated 👍

File: docs/pages/machine_learning.rst (hunk 18 above)
Comment (rajagurunath) [reply docs/pages/machine_learning.rst]: updated 👍

File: dask_sql/physical/rel/custom/export_model.py (hunk 19 above)
Comment (rajagurunath) [reply dask_sql/physical/rel/custom/export_model.py]: got it now changed the `RuntimeError` error into `ImportError`

File: docs/pages/machine_learning.rst (hunk 25 above)
Comment (rajagurunath) [reply docs/pages/machine_learning.rst:119]: yeah it's really looking nice this way, changed in most of the places. 👍

File: planner/src/main/codegen/includes/model.ftl (hunk 26 above)
Comment (rajagurunath) [reply planner/src/main/codegen/includes/model.ftl]: got it, removed

File: tests/integration/test_model.py (hunk 27 above)
Comment (rajagurunath) [reply tests/integration/test_model.py:344]: learned from you, Thanks ❤️

File: tests/integration/test_model.py (hunk 28 above)
Comment (rajagurunath) [reply tests/integration/test_model.py]: makes sense 👍

File: tests/integration/test_model.py (hunk 29 above)
Comment (rajagurunath) [reply tests/integration/test_model.py]: Amazing, this is new to me, thanks

**Comment by rajagurunath** [comment]
Hi @nils-braun,

Thanks a lot for these comments, I really enjoyed doing this, learning a lot from you each time, Amazed by your clear-cut explanation in each comment, and once again thanks for the time spent on giving these comments really fascinating the level of details each comment provides. ❤️ 

And Regarding the test checks failing in the workflow, got to know that the following steps are failing:
1. while installing sqlalchemy and docker, building wheel for package `ciso8601`
2. And while uploading the pytest test result, getting this error path not found `/home/runner/work/dask-sql/dask-sql/junit/test-results.xml` 

Haven't had much experience in GitHub actions, not sure how to debug this, may need your help/input here.

Thanks

**Comment by nils-braun** [comment]
Thanks for your updates, looks quite good to me now. Just the failing build...
Your observation is completely correct, for whatever reasons the pip-installation of the ciso8601 package is failing.
Ok, I will try to fix this in a different branch and then merge back to main. I hope this is somewhat easy to fix :-/

**Comment by nils-braun** [comment]
@rajagurunath - should be fixed now! If you want, please merge in the newest main branch commits and your PR should also build successfully!

**Comment by rajagurunath** [comment]
sure, Thanks a lot @nils-braun  for the fix

**Comment by rajagurunath** [comment]
Thanks to Code coverage, Identified a bug in the code.

export_model.py checks model instances of the different frameworks with their respective classes like xgb.core.Booster, but this type of xgboost booster was not sklearn compatible (doesn't have fit-predict API's) which are trained differently as [here](https://github.com/mlflow/mlflow/blob/master/examples/xgboost/train.py)

I believe since this is not sklearn compatible which will be trained in a different manner which is currently not implemented in dask-sql, please correct me if I am wrong here. so as a remedy I have added only one check which checks whether the model is sklearn compatible or not if it is compatible it will be serialized with mlflow.sklearn flavor and which is working out of the box for xgboost (XGBClassifier), lightgbm(LGBMClassifier), etc. 

made some changes and pushed again, please review the changes and let me know if anything needs to be added. Apologies for not testing this end-to-end earlier.

Thanks

**Comment by nils-braun** [comment]
Perfect, thanks for having a look!
The coverage is still not 100%, probably because the mlflow test is skipped (as it is not installed on the GitHub runner).
Would you think it makes sense to install it? In this case, could you add it to the GitHub `test.yml` file?

**Comment by nils-braun** [comment]
Green :-)
Lgtm!
Thanks again for your nice work
